### PR TITLE
Draft design for the matrix context?

### PR DESCRIPTION
**Pull request by jmfayard**
See #368  and #297

This is just a draft of what the API looks like.
I didn't try to generate the YAML correctly, and there are some features of the matrix that I don't really understand, like having objects, include (not supported) and excludes

wdyt?

### CODE REVIEW

File: library/src/test/kotlin/it/krzeminski/githubactions/dsl/expressions/MatrixPlayground.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            "node" to listOf("14", "16")
+        )
+    }
+
+    val simpleWorkflowWithMatrix = workflow(
+        name = "Test matrix",
+        on = listOf(Push()),
+        sourceFile = Path.of("some_workflow.main.kts"),
+    ) {
+        val matrix = object : StrategyMatrix() {
```
Comment (krzema12) [inline library/src/test/kotlin/it/krzeminski/githubactions/dsl/expressions/MatrixPlayground.kt:30]: Have you considered making the strategy one of `workflow`'s arguments, like it's done now in the simplified implementation?

File: library/src/main/kotlin/it/krzeminski/githubactions/dsl/expressions/contexts/MatrixContext.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    public fun strings(vararg value: String): Property {
+        return Property(value.toList())
+    }
+
+    public fun integers(vararg value: Int): Property {
+        return Property(value.map { it.toString() })
+    }
+
+    public fun objects(vararg value: Map<String, String>): Property {
```
Comment (krzema12) [inline library/src/main/kotlin/it/krzeminski/githubactions/dsl/expressions/contexts/MatrixContext.kt:21]: The objects can have also non-string attribute values.

**Comment by krzema12** [comment]
@jmfayard thanks for starting working on it!

IMO first we should enumerate all features of GitHub strategies, understand them well (with some examples in YAML), decide which ones we want to implement (let's start with "all"), and then propose examples how each of them is covered in the DSL. Does it make sense?

File: library/src/test/kotlin/it/krzeminski/githubactions/dsl/expressions/MatrixPlayground.kt (hunk 1 above)
Comment (jmfayard) [reply library/src/test/kotlin/it/krzeminski/githubactions/dsl/expressions/MatrixPlayground.kt:30]: to be honest, I am not a fan that `workflow()`, `job()` have like 10 parameters each

**Comment by jmfayard** [comment]
yes absolutely, that's also what I had in moind.
Do a first proof of concept of what a type-safe api could look like,
then go back to understand the corner cases of what the strategy matrix supports

File: library/src/test/kotlin/it/krzeminski/githubactions/dsl/expressions/MatrixPlayground.kt (hunk 1 above)
Comment (krzema12) [reply library/src/test/kotlin/it/krzeminski/githubactions/dsl/expressions/MatrixPlayground.kt:30]: These approaches have both pros and cons. Some of the pros of the argument approach I see:
* better discoverability in the IDE - the users are more likely to discover an argument than some extra DSL piece
* compile-time enforcement of having this matrix defined just once in a well-defined, canonical place

One of the cons is sure the number of arguments which starts to become hard to maintain.

Let's defer this decision until we have a better understanding on the features we want to implement.

**Comment by krzema12** [comment]
@jmfayard I'd also love to engage @Vampire into this design, so that he can play with some intermediate version of the API in the context of his workflows (some of them make heavy use of the matrices).

**Comment by krzema12** [comment]
@jmfayard are you planning to work on this, or you're fine with me taking over the task? I'm drafting the API in https://github.com/krzema12/github-workflows-kt/pull/635

**Comment by krzema12** [comment]
Please reopen if needed.
